### PR TITLE
Improvements to the way failures are handled and state saving

### PR DESCRIPTION
**Pull request by Luke2Sky**
This pull request aims to improve how unexpected process stopping  in multi_component_hydrogen_bond_propensity_report.py loses all progress made. This happens when the computer in which it is being run is unexpectedly shutdown or when the script uses too much memory.  I also listed the coformers for which there is a failure in the file failures.txt. This improvement is not enough for a new script based in MCHBP.  This was tested with HXACAN28 and 3 coformers that we are using in LaMuCrEs-IFSC-USP. It also worked in a longer test with 90 coformers for a specific molecule we are working with.

### CODE REVIEW

File: scripts/multi_component_hydrogen_bond_propensity/multi_component_hydrogen_bond_propensity_report.py [hunk 1]
```diff
@@ -343,22 +345,35 @@ def main(structure, work_directory, library, csdrefcode):
...
+                print("Propensity calculation failure for %s!" % coformer_name)
+                mc_dictionary[coformer_name] = ["N/A", "N/A", "N/A", "N/A", "N/A", crystal.identifier]
+                failures.append(coformer_name)
 
     # Make sense of the outputs of all the calculations
     mc_hbp_screen = sorted(mc_dictionary.items(), key=lambda e: 0 if e[1][0] == 'N/A' else e[1][0], reverse=True)
     diagram_file = make_diagram(api_molecule, work_directory)
     chart_file = make_mc_chart(mc_hbp_screen, directory, api_molecule)
     make_mc_report(structure, mc_hbp_screen, work_directory, diagram_file, chart_file)
+    with open("failures.txt", 'w') as file:
```
Comment (Alex-AMC) [inline scripts/multi_component_hydrogen_bond_propensity/multi_component_hydrogen_bond_propensity_report.py:372]: I think this should have the same output location as the above with the 
```python
os.path.join(directory,...) 
```

File: scripts/multi_component_hydrogen_bond_propensity/multi_component_hydrogen_bond_propensity_report.py [hunk 2]
```diff
@@ -343,22 +345,35 @@ def main(structure, work_directory, library, csdrefcode):
...
-            pair_output(crystal.identifier, propensities, donors, acceptors, coordination_scores, directory)
-            mc_dictionary[coformer_name] = get_mc_scores(propensities, crystal.identifier)
-
-        except RuntimeError:
-            print("Propensity calculation failure for %s!" % coformer_name)
-            mc_dictionary[coformer_name] = ["N/A", "N/A", "N/A", "N/A", "N/A", crystal.identifier]
+        if os.path.exists(os.path.join(directory, "success.json")):
+            with open(os.path.join(directory, "success.json"), "r") as file:
+                tloaded = json.load(file)
+            mc_dictionary[coformer_name] = tloaded
```
Comment (Alex-AMC) [inline scripts/multi_component_hydrogen_bond_propensity/multi_component_hydrogen_bond_propensity_report.py:351]: Nice addition :)

**Review by Alex-AMC — CHANGES_REQUESTED** [review]
Looks good. I'd recommend implementing the change suggested and then we can look at getting it merged in.

**Comment by Luke2Sky** [comment]
I already implemented it. The first commit is for one of the 2 second is for the second.

File: scripts/multi_component_hydrogen_bond_propensity/multi_component_hydrogen_bond_propensity_report.py (hunk 1 above)
Comment (Luke2Sky) [reply scripts/multi_component_hydrogen_bond_propensity/multi_component_hydrogen_bond_propensity_report.py:372]: I have libraries of coformers with 90+ elements. Looking at each one individually to verify if it failed would be very tiresome. That's why I did it outside the directory of each coformer. That file lists the failures and then I can look inside the failed dir.

File: scripts/multi_component_hydrogen_bond_propensity/multi_component_hydrogen_bond_propensity_report.py (hunk 2 above)
Comment (Luke2Sky) [reply scripts/multi_component_hydrogen_bond_propensity/multi_component_hydrogen_bond_propensity_report.py:351]: Thank you!

**Review by pmbulit — CHANGES_REQUESTED** [review]

**Review by pmbulit — APPROVED** [review]
Looks good to me. Thanks for the changes!

**Review by Alex-AMC — APPROVED** [review]

**Comment by Luke2Sky** [comment]
I didn't get the branch check. Did I do something wrong?

**Comment by Alex-AMC** [comment]
@Luke2Sky don't worry. This is something we will be changing soon, but for now, we will try to merge the changes as they are :)

**Pull request merged**